File: stringsArrays/SearchInsertPos.java
/**
 * Search insert position (136):
 * Given a sorted array and a target value, return the index if the target is found. If not, return the index where it would be if it were inserted in order. You may assume no duplicates in the array.
 * Here are few examples.
 * [1,3,5,6], 5 -> 2
 * [1,3,5,6], 2 -> 1
 * [1,3,5,6], 7 -> 4
 * [1,3,5,6], 0 -> 0
 * http://www.programcreek.com/2013/01/leetcode-search-insert-position/
 */

/*
 * time: O(log n)
 * space: O(1)
 */

import java.util.Arrays;

public class SearchInsertPos {
    public int searchInsert(int[] nums, int target) {
        if (nums.length == 0) {
            return 0;
        }

        int start = 0, end = nums.length - 1, mid = 0;
        while (start <= end) {
            mid = (start + end) / 2;
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] < target) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }

        return nums[mid] > target ? mid : mid + 1;
    }

    public static void main(String[] args) {
        if (args.length < 3) {
            System.out.println("Usage: java <prog> <s> <size> <item1> <item2> ...");
            System.exit(1);
        }

        int s = Integer.parseInt(args[0]);
        int size = Integer.parseInt(args[1]);
        int[] arr = new int[size];
        for (int i = 0; i < size; ++i) {
            arr[i] = Integer.parseInt(args[i + 2]);
        }
        System.out.println("target: " + s);
        System.out.println("nums: " + Arrays.toString(arr));

        SearchInsertPos sip = new SearchInsertPos();
        int result = sip.searchInsert(arr, s);
        System.out.println("result: " + result);
    }
}

File: searchInsertPos.cpp
/* Search insert position (136):
Given a sorted array and a target value, return the index if the target is found. 
If not, return the index where it would be if it were inserted in order. 
You may assume no duplicates in the array. 
Here are few examples.
[1,3,5,6], 5 -> 2
[1,3,5,6], 2 -> 1
[1,3,5,6], 7 -> 4
[1,3,5,6], 0 -> 0
http://www.programcreek.com/2013/01/leetcode-search-insert-position/ */

#include <iostream>
#include <vector>

using namespace std;

int getIndex(const vector<int> &a, int target, int start, int end)
{
	int mid = (start + end) / 2;
	
	if (a[mid] == target) {
		return mid;
	} else if (a[mid] < target) {
		if ((mid + 1) > end || ((mid + 1) <= end && a[mid + 1] > target)) {
			return mid + 1;
		}
		
		return getIndex(a, target, mid + 1, end);
	} else {
		if ((mid - 1) < start || ((mid - 1) >= start && a[mid - 1] < target)) {
			return mid;
		}
		
		return getIndex(a, target, start, mid - 1);
	}
}

int main(int argc, char **argv)
{
	if (argc < 4) {
		cout << "Usage: a.out <target> <arr size> <arr>" << endl;
		return 1;
	}
	
	int target = atoi(argv[1]);
	int aSize = atoi(argv[2]);
	vector<int> a;
	for (int i = 3; i < 3 + aSize; ++i) {
		a.push_back(atoi(argv[i]));
	}
	
	int idx = getIndex(a, target, 0, a.size() - 1);
	cout << "Index: " << idx << endl;
	
	return 0;
}